crates/contracts/solidity/Solver.sol

```diff
@@ -0,0 +1,84 @@
+/// estimation.
+contract Solver {
+    using Caller for *;
+    using Math for *;
+    using SafeERC20 for *;
+
+    uint256 private _simulationOverhead;
+    uint256[] private _queriedBalances;
+
+    /// @dev Executes the given transaction from the context of a solver.
+    /// That way we don't have to fake the authentication logic of the
+    /// settlement contract as this address should actually be a
+    /// verified solver.
+    ///
+    /// @param trader - address of the order owner doing the trade
+    /// @param sellToken - address of the token being sold
+    /// @param sellAmount - amount being sold
+    /// @param nativeToken - ERC20 version of the chain's token
+    /// @param receiver - address receiving the bought tokens
+    /// @param settlementCall - the calldata of the `settle()` call
+    ///
+    /// @return gasUsed - gas used for the `settle()` call
+    /// @return queriedBalances - list of balances stored during the simulation
+    function swap(
+        address payable trader,
+        address sellToken,
+        uint256 sellAmount,
+        address nativeToken,
+        address payable receiver,
+        bytes calldata settlementCall
+    ) external returns (
+        uint256 gasUsed,
+        uint256[] memory queriedBalances
+    ) {
+        require(msg.sender == address(this), "only simulation logic is allowed to call 'swap' function");
+        // Prepare the trade in the context of the trader so we are allowed
+        // to set approvals and things like that.
+        Trader(trader).prepareSwap(sellToken, sellAmount, nativeToken, receiver);
+        uint256 gasStart = gasleft();
+        // TODO can we assume the overhead of this function call to be negligible due to inlining?
+        address(SETTLEMENT).doCall(settlementCall);
+        gasUsed = gasStart - gasleft() - _simulationOverhead;
+        queriedBalances = _queriedBalances;
+    }
+
+    /// @dev Helper function that reads the `owner`s balance for a given `token` and
+    /// stores it. These stored balances will be returned as part of the simulation
+    /// `Summary`.
+    /// @param token - which token's we read the balance from
+    /// @param owner - whos balance we are reading
+    function storeBalance(address token, address owner) external {
+        uint256 gasStart = gasleft();
+        if (token == 0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE) {
+            // native ETH
+            _queriedBalances.push(owner.balance);
+        } else {
+            // regular ERC20 token
+            _queriedBalances.push(IERC20(token).balanceOf(owner));
+        }
+        // Account for costs of gas used outside of metered section.
+        uint256 measurementOverhead;
+        if (_queriedBalances.length == 1) {
+            // reading cold storage costs more which results in higher overhead for the first call
+            measurementOverhead = 13355;
+        } else {
+            measurementOverhead = 4460;
+        }
```

Comment:
You can make this slightly more succinct with:
```suggestion
        _queriedBalances.push(
            token == 0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE
                ? owner.balance
                : IERC20(token).balanceOf(owner)
        );
        // Account for costs of gas used outside of metered section. Noting that
        // reading cold storage costs more which results in higher overhead for the first call
        uint256 measurementOverhead = _queriedBalances.length == 1 ? 13355 : 4460;
```

Personally, I kind of prefer the `if/else` for the first one as I find it easier to read, but for the second one (since they are simple constants) I find the ternary operator a bit nicer.

Fine as is though!